### DISCUSSION TITLE
Coding Best Practices

### DISCUSSION THREAD

**Discussion by sub1120** [lmsGenie/lms-app#41] [Ideas]
Here we can discuss and decide coding best practices for better code quality.

**Comment by sub1120** [comment]
## Icon Library Management

- How we are going to manage icons?
- Where we will store icons? ( /assets/icons, /public/icons, or CDN)
- In which format we will store them? (png, jpg, svg etc.)?
- How we will handle the variations in icons (eg. change color in dark mode)? See one example below, how the color of icon changes on hover,
![image](https://github.com/lmsGenie/client/assets/43786036/dc310f5c-106a-4733-8417-b29634b4a724)


Here is one article, but we still need more clarity.
https://forum.figma.com/t/every-icon-best-practice-you-need-to-know/6445

**Comment by sub1120** [comment]
@shivamvijaywargi , please read this article. https://benadam.me/thoughts/react-svg-sprites/

I think we can go with 3rd logic, "Rendering Icons using SVG Sprites". It solves 3 problems,
1. Loading time
2. Storing is easy
3. Change Icon style with <use> tag (https://developer.mozilla.org/en-US/docs/Web/SVG/Element/use)

**Reply by shivamvijaywargi** [reply]
Won't it be a little hard?

**Reply by sub1120** [reply]
Why bro? 

We Only need to setup one time.

**Reply by shivamvijaywargi** [reply]
Okay, let's see.

**Comment by sub1120** [comment]
## How to use?
1. Import Icon component
```javascript
import Icon from "@/components/icon/Icon";
```

2. Next we just need to use `<Icon/>` component 
```javascript
<Icon id="icon-id"/>
```
where `icon-id`, can be from any of one valid types suggested by the editor.

![image](https://github.com/lmsGenie/client/assets/43786036/803316e7-3c8c-4ccc-92c0-89b71e708522)

Additionally, we can alter icon size, fill color, and stroke color, using tailwind classes
```javascript
<Icon id="MapPin" className="h-12 w-12 hover:stroke-primary"></Icon>
```
## How to add a new Icon
1. Add new `<symbol>`.
```javascript
// /public/sprite.svg
<symbol id="icon-id" viewBox="0 0 24 24">
     // Your icon paths
</symbol>
```
**Note:** Remove following attributes from your paths, if present fillColor="anyValue" stroke="currentColor", for better control over style overrding.

2. Add your `icon-id` to `idType`

```javascript
// types/icon/index.ts
export type idType = "icon-id" | .....
```
This step is to enhance code suggestions, while using Icons.


More Examples
```javascript
// plain icons with default fill, stroke, and size.
 <Icon id="XCircle"></Icon>
 <Icon id="XCircle"></Icon>
 <Icon id="XCircle"></Icon>
 <Icon id="XCircle"></Icon>
 <Icon id="XCircle"></Icon>
 <Icon id="XCircle"></Icon>

// change default stroke color
<Icon id="Pen" className="stroke-primary"></Icon>
<Icon id="Pen" className="stroke-primary"></Icon>
<Icon id="Pen" className="stroke-primary"></Icon>
<Icon id="Pen" className="stroke-primary"></Icon>
<Icon id="Pen" className="stroke-primary"></Icon>
<Icon id="Pen" className="stroke-primary"></Icon>

// fill color with no stroke
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>
<Icon id="ThumbsUp" className="fill-primary stroke-none"></Icon>

// stroke color success
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>
<Icon id="CheckSquare" className="stroke-success-600  "></Icon>

// stroke color warning
<Icon id="Prohibit" className="stroke-warning-600"></Icon>
<Icon id="Prohibit" className="stroke-warning-600"></Icon>
<Icon id="Prohibit" className="stroke-warning-600"></Icon>
<Icon id="Prohibit" className="stroke-warning-600"></Icon>
<Icon id="Prohibit" className="stroke-warning-600"></Icon>
<Icon id="Prohibit" className="stroke-warning-600"></Icon>

// size variation + hover effect
<Icon id="MapPin" className="hover:stroke-primary"></Icon>
<Icon id="MapPin" className="h-12 w-12 hover:stroke-primary"></Icon>
<Icon id="MapPin" className="h-14 w-14 hover:stroke-primary"></Icon>
<Icon id="MapPin" className="h-16 w-16 hover:stroke-primary"></Icon>
<Icon id="MapPin" className="h-20 w-20 hover:stroke-primary"></Icon>
<Icon id="MapPin" className="h-24 w-24 hover:stroke-primary"></Icon>
```
Output
![image](https://github.com/lmsGenie/client/assets/43786036/1a89f7f6-9d12-4118-96fa-b4ad1f0bb14c)
